File: engines/lua/g2048/heuristic.lua
-- AI.md §2: integer heuristic, per-line tables, corner term, breakdown.
--
-- Line scores are stored as Lua integers: every value is an exact integer
-- below 2^53, so summing integers and converting once gives the same float64
-- as the reference's float additions.

local bitboard = require("g2048.bitboard")

local M = {}

M.WEIGHT_KEYS = { "lost", "empty", "merges", "mono", "sum", "smooth", "stable", "corner" }
M.FEATURE_KEYS = { "empty", "merges", "mono", "sum", "smooth", "stable" }

function M.heuristic_v1()
  return { lost = 200000, empty = 270, merges = 700, mono = 47, sum = 11, smooth = 0, stable = 0, corner = 0 }
end

function M.copy_weights(w)
  local c = {}
  for _, k in ipairs(M.WEIGHT_KEYS) do c[k] = w[k] or 0 end
  return c
end

function M.weights_equal(a, b)
  for _, k in ipairs(M.WEIGHT_KEYS) do
    if (a[k] or 0) ~= (b[k] or 0) then return false end
  end
  return true
end

--- Features of a 16-bit line value.
function M.features(v)
  local r = { v & 0xf, (v >> 4) & 0xf, (v >> 8) & 0xf, (v >> 12) & 0xf }
  local empty, merges, sum, smooth, stable = 0, 0, 0, 0, 0
  local prev, counter = 0, 0
  for i = 1, 4 do
    local rank = r[i]
    sum = sum + rank * rank * rank
    if rank == 0 then
      empty = empty + 1
    else
      if prev == rank then
        counter = counter + 1
      elseif counter > 0 then
        merges = merges + 1 + counter
        counter = 0
      end
      prev = rank
    end
  end
  if counter > 0 then merges = merges + 1 + counter end
  local monoL, monoR = 0, 0
  for i = 2, 4 do
    local p, q = r[i - 1], r[i]
    local a, b = p * p * p * p, q * q * q * q
    if r[i - 1] > r[i] then
      monoL = monoL + a - b
    else
      monoR = monoR + b - a
    end
    if r[i - 1] ~= 0 and r[i] ~= 0 then
      smooth = smooth + math.abs(r[i - 1] - r[i])
    end
  end
  if empty == 0 and (monoL == 0 or monoR == 0) then stable = 1 end
  return {
    empty = empty,
    merges = merges,
    mono = math.min(monoL, monoR),
    sum = sum,
    smooth = smooth,
    stable = stable,
  }
end

--- Weighted line score (integer).
function M.line_score(f, w)
  return w.lost + w.empty * f.empty + w.merges * f.merges - w.mono * f.mono - w.sum * f.sum
    - w.smooth * f.smooth + w.stable * f.stable
end

local cache = {}

local function weights_key(w)
  local parts = {}
  for i, k in ipairs(M.WEIGHT_KEYS) do parts[i] = tostring(w[k] or 0) end
  return table.concat(parts, ",")
end

--- Memoised 65536-entry line-score table (integer values, index 0..65535).
function M.line_table(w)
  local key = weights_key(w)
  local t = cache[key]
  if t then return t end
  t = {}
  local fw = M.copy_weights(w)
  for v = 0, 65535 do
    t[v] = M.line_score(M.features(v), fw)
  end
  cache[key] = t
  return t
end

--- W.corner * maxRank if a corner holds the max rank, else 0 (integer).
function M.corner_term(b, corner_weight)
  if corner_weight == 0 then return 0 end
  local mx = 0
  for i = 0, 15 do
    local n = (b >> (4 * i)) & 0xf
    if n > mx then mx = n end
  end
  if (b & 0xf) == mx or ((b >> 12) & 0xf) == mx or ((b >> 48) & 0xf) == mx or (b >> 60) == mx then
    return corner_weight * mx
  end
  return 0
end

--- Heuristic value of a bitboard as a float.
function M.evaluate(b, t, corner_weight)
  local x = bitboard.transpose(b)
  local v = t[b & 0xffff] + t[(b >> 16) & 0xffff] + t[(b >> 32) & 0xffff] + t[b >> 48]
    + t[x & 0xffff] + t[(x >> 16) & 0xffff] + t[(x >> 32) & 0xffff] + t[x >> 48]
  if corner_weight ~= 0 then v = v + M.corner_term(b, corner_weight) end
  return v + 0.0
end

--- Heuristic decomposition summed over all 8 lines.
function M.breakdown(b, w)
  local x = bitboard.transpose(b)
  local lines = {
    b & 0xffff, (b >> 16) & 0xffff, (b >> 32) & 0xffff, b >> 48,
    x & 0xffff, (x >> 16) & 0xffff, (x >> 32) & 0xffff, x >> 48,
  }
  local acc = { empty = 0, merges = 0, mono = 0, sum = 0, smooth = 0, stable = 0 }
  for _, v in ipairs(lines) do
    local f = M.features(v)
    for _, k in ipairs(M.FEATURE_KEYS) do acc[k] = acc[k] + f[k] end
  end
  local corner = w.corner or 0
  acc.corner = M.corner_term(b, corner) + 0.0
  acc.total = M.evaluate(b, M.line_table(w), corner)
  return acc
end

return M

File: engines/java/src/main/java/dev/g2048/ai/Heuristic.java
package dev.g2048.ai;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/** Integer heuristic (AI.md §2). */
public final class Heuristic {
  private Heuristic() {}

  /** Integer heuristic weights. */
  public record Weights(long lost, long empty, long merges, long mono, long sum, long smooth, long stable, long corner) {
    public Map<String, Object> toMap() {
      var m = new LinkedHashMap<String, Object>();
      m.put("lost", lost);
      m.put("empty", empty);
      m.put("merges", merges);
      m.put("mono", mono);
      m.put("sum", sum);
      m.put("smooth", smooth);
      m.put("stable", stable);
      m.put("corner", corner);
      return m;
    }

    public Weights with(String key, long v) {
      return switch (key) {
        case "lost" -> new Weights(v, empty, merges, mono, sum, smooth, stable, corner);
        case "empty" -> new Weights(lost, v, merges, mono, sum, smooth, stable, corner);
        case "merges" -> new Weights(lost, empty, v, mono, sum, smooth, stable, corner);
        case "mono" -> new Weights(lost, empty, merges, v, sum, smooth, stable, corner);
        case "sum" -> new Weights(lost, empty, merges, mono, v, smooth, stable, corner);
        case "smooth" -> new Weights(lost, empty, merges, mono, sum, v, stable, corner);
        case "stable" -> new Weights(lost, empty, merges, mono, sum, smooth, v, corner);
        case "corner" -> new Weights(lost, empty, merges, mono, sum, smooth, stable, v);
        default -> this;
      };
    }

    public static Weights fromMap(Map<String, Object> m) {
      Weights w = new Weights(0, 0, 0, 0, 0, 0, 0, 0);
      for (var e : m.entrySet()) w = w.with(e.getKey(), dev.g2048.json.Json.lng(e.getValue()));
      return w;
    }
  }

  /** Canonical profile HEURISTIC_V1. */
  public static final Weights HEURISTIC_V1 = new Weights(200000, 270, 700, 47, 11, 0, 0, 0);

  /** Per-line features. */
  public record LineFeatures(long empty, long merges, long mono, long sum, long smooth, long stable) {
    public LineFeatures plus(LineFeatures o) {
      return new LineFeatures(
          empty + o.empty, merges + o.merges, mono + o.mono, sum + o.sum, smooth + o.smooth, stable + o.stable);
    }

    public Map<String, Object> toMap() {
      var m = new LinkedHashMap<String, Object>();
      m.put("empty", empty);
      m.put("merges", merges);
      m.put("mono", mono);
      m.put("sum", sum);
      m.put("smooth", smooth);
      m.put("stable", stable);
      return m;
    }
  }

  public static LineFeatures features(int v) {
    long[] r = {v & 0xf, (v >>> 4) & 0xf, (v >>> 8) & 0xf, (v >>> 12) & 0xf};
    long empty = 0, merges = 0, sum = 0, smooth = 0, stable = 0;
    long prev = 0, counter = 0;
    for (long rank : r) {
      sum += rank * rank * rank;
      if (rank == 0) {
        empty++;
        continue;
      }
      if (prev == rank) counter++;
      else if (counter > 0) {
        merges += 1 + counter;
        counter = 0;
      }
      prev = rank;
    }
    if (counter > 0) merges += 1 + counter;
    long monoL = 0, monoR = 0;
    for (int i = 1; i < 4; i++) {
      long a = r[i - 1] * r[i - 1] * r[i - 1] * r[i - 1];
      long b = r[i] * r[i] * r[i] * r[i];
      if (r[i - 1] > r[i]) monoL += a - b;
      else monoR += b - a;
      if (r[i - 1] != 0 && r[i] != 0) smooth += Math.abs(r[i - 1] - r[i]);
    }
    if (empty == 0 && (monoL == 0 || monoR == 0)) stable = 1;
    return new LineFeatures(empty, merges, Math.min(monoL, monoR), sum, smooth, stable);
  }

  public static long lineScore(LineFeatures f, Weights w) {
    return w.lost() + w.empty() * f.empty() + w.merges() * f.merges() - w.mono() * f.mono() - w.sum() * f.sum()
        - w.smooth() * f.smooth() + w.stable() * f.stable();
  }

  private static final Map<Weights, double[]> TABLES = new ConcurrentHashMap<>();

  /** Memoised 65536-entry per-line score table. */
  public static double[] lineTable(Weights w) {
    return TABLES.computeIfAbsent(w, k -> {
      double[] t = new double[65536];
      for (int v = 0; v < 65536; v++) t[v] = (double) lineScore(features(v), k);
      return t;
    });
  }

  public static double cornerTerm(long b, long cornerWeight) {
    if (cornerWeight == 0) return 0;
    long mx = 0;
    for (int i = 0; i < 16; i++) {
      long n = (b >>> (4 * i)) & 0xf;
      if (n > mx) mx = n;
    }
    if ((b & 0xf) == mx || ((b >>> 12) & 0xf) == mx || ((b >>> 48) & 0xf) == mx || (b >>> 60) == mx)
      return (double) (cornerWeight * mx);
    return 0;
  }

  public static double evaluate(long b, double[] t, long cornerWeight) {
    long c = Bitboard.transpose(b); // rows of c are the columns of b, in order
    double v = t[(int) (b & 0xffff)] + t[(int) ((b >>> 16) & 0xffff)] + t[(int) ((b >>> 32) & 0xffff)]
        + t[(int) (b >>> 48)] + t[(int) (c & 0xffff)] + t[(int) ((c >>> 16) & 0xffff)]
        + t[(int) ((c >>> 32) & 0xffff)] + t[(int) (c >>> 48)];
    if (cornerWeight != 0) v += cornerTerm(b, cornerWeight);
    return v;
  }

  /** Heuristic decomposition summed over all 8 lines. */
  public record Breakdown(LineFeatures features, double corner, double total) {
    public Map<String, Object> toMap() {
      var m = features.toMap();
      m.put("corner", corner);
      m.put("total", total);
      return m;
    }
  }

  public static Breakdown breakdown(long b, Weights w) {
    int[] lines = {
      (int) (b & 0xffff), (int) ((b >>> 16) & 0xffff), (int) ((b >>> 32) & 0xffff), (int) (b >>> 48),
      Bitboard.column(b, 0), Bitboard.column(b, 1), Bitboard.column(b, 2), Bitboard.column(b, 3)
    };
    LineFeatures acc = new LineFeatures(0, 0, 0, 0, 0, 0);
    for (int v : lines) acc = acc.plus(features(v));
    return new Breakdown(acc, cornerTerm(b, w.corner()), evaluate(b, lineTable(w), w.corner()));
  }
}
